gmsh_utils/math_utils.py

```diff
@@ -0,0 +1,177 @@
+
+        # Shift projected_polygon by one point to the left (circular shift)
+        shifted_polygon = np.roll(projected_polygon, -1, axis=0)
+
+        # Unpack coordinates
+        p1eta, p1xi = projected_polygon[:, 0], projected_polygon[:, 1]
+        p2eta, p2xi = shifted_polygon[:, 0], shifted_polygon[:, 1]
+
+        # Check if the point is within the xi-bounds of the polygon edges
+        between_xi_bounds = ((p1xi < xi + eps) & (xi < p2xi + eps)) | ((p2xi < xi + eps) & (xi < p1xi + eps))
+
+        # Compute the intersection points of the polygon edges with the horizontal line at y
+        eta_intersections = (p2eta - p1eta) * (xi - p1xi) / (p2xi - p1xi) + p1eta
+
+        # check if point is on the edge
+        on_edge = np.isclose(eta, eta_intersections) & between_xi_bounds
+        if on_edge.any():
+            return True
+
+        # Check if the point is to the left of the intersection points
+        left_of_intersection = between_xi_bounds & (eta < eta_intersections+eps)
+
+        # point is inside if an odd number of edges are to the right of the point
+        is_point_inside: bool = np.any(left_of_intersection) & np.sum(left_of_intersection) % 2 == 1
+        return is_point_inside
+
+    @staticmethod
+    def is_point_on_plane(point: npt.NDArray[np.float64],
+                          plane_point: npt.NDArray[np.float64],
+                          plane_normal: npt.NDArray[np.float64]) -> bool:
+        """
+        Check if a point is on a plane.
+
+        Args:
+            - point (npt.NDArray[np.float64]): Point to check.
+            - plane_point (npt.NDArray[np.float64]): Point on the plane.
+            - plane_normal (npt.NDArray[np.float64]): Normal of the plane.
+
+        Returns:
+            - bool: True if the point is on the plane, False
+
+
+        """
+
+        # check if the point is on the plane
+        is_point_on_plane: bool = np.isclose(np.dot(point - plane_point, plane_normal), 0.0)
+        return is_point_on_plane
+
+    @staticmethod
+    def calculate_normal_plane(plane_vertices: npt.NDArray[np.float64]) -> npt.NDArray[np.float64]:
+        """
+        Calculate the normal of a plane defined by three vertices.
+
+        Args:
+            - plane_vertices (npt.NDArray[np.float64]): Vertices of the plane.
+
+        Raises:
+            - ValueError: If the vertices are collinear.
+
+        Returns:
+            - npt.NDArray[np.float64]: Normal of the plane.
+        """
+
+        v1 = plane_vertices[1] - plane_vertices[0]
+
+        # make sure that the vertices are not collinear
+        for i in range(2, len(plane_vertices)):
+            v2 = plane_vertices[i] - plane_vertices[0]
+            normal: npt.NDArray[np.float64] = np.cross(v1, v2)
+
+            # return the normal if it is not zero
+            if not np.allclose(normal, 0):
+                return normal / np.linalg.norm(normal)
+
+        raise ValueError("All plane vertices are collinear.")
```

Comment:
This is only for 3 nodes? if so, no need for the for loop and this should work: 
    v1 = plane_vertices[1] - plane_vertices[0]
    v2 = plane_vertices[i] - plane_vertices[0]
    normal_vector = np.cross(v1, v2)
    normal_vector_normalized = normal_vector / np.linalg.norm(normal_vector)
    
    if np.all(normal_vector == 0):
       ValueError

Reply:
no the docstring is wrong, it should work for any polygon